File: web/spa/src/lib/events.ts
/**
 * The push channel: exactly one `EventSource` for the whole app.
 *
 * One stream, a registry of per-name handlers, and a 2-second reconnect (R-FE-7,
 * Inventory §3). "Exactly one" is a rule rather than an optimisation: a component that
 * opens its own stream doubles every event, and the duplicate arrives at a *different*
 * moment — so bugs caused by it look like race conditions rather than like duplication.
 *
 * Authentication is the session cookie, which the browser attaches automatically.
 * `EventSource` cannot send headers, and that constraint is the reason the session is a
 * cookie at all (D22).
 *
 * Two consumer-side rules that live where the data lands rather than here, noted because
 * they are easy to miss:
 *
 * - `job.updated` may be a **full row or a partial** `{id, result}` during bulk progress.
 *   Merge by id (Inventory §10.30) — a 240-item retag would otherwise re-send the whole
 *   row on every tick.
 * - `item.created` must be **ignored by the library grid while any filter or search is
 *   active** (Inventory §10.25). Prepending a card that does not match the active filter
 *   is a lie about what the user is looking at.
 */

/** The event names the server publishes (R-BE-15). */
export type EventName =
  | "item.created"
  | "item.updated"
  | "item.deleted"
  | "project.detected"
  | "project.updated"
  | "project.removed"
  | "job.updated"
  | "vocabulary.updated";

export const EVENT_NAMES: readonly EventName[] = [
  "item.created",
  "item.updated",
  "item.deleted",
  "project.detected",
  "project.updated",
  "project.removed",
  "job.updated",
  "vocabulary.updated",
] as const;

type Handler = (payload: unknown) => void;

const RECONNECT_MS = 2_000;

/** The single shared stream. */
export class EventStream {
  #source: EventSource | null = null;
  #handlers = new Map<EventName, Set<Handler>>();
  #reconnect: ReturnType<typeof setTimeout> | null = null;

  /** Subscribe to one event name. Returns the unsubscribe function. */
  on(name: EventName, handler: Handler): () => void {
    let handlers = this.#handlers.get(name);
    if (!handlers) {
      handlers = new Set();
      this.#handlers.set(name, handlers);
    }
    handlers.add(handler);
    return () => {
      handlers?.delete(handler);
    };
  }

  /** Open the stream. Idempotent. */
  connect(): void {
    if (this.#source) return;

    const source = new EventSource("/api/events");
    this.#source = source;

    for (const name of EVENT_NAMES) {
      source.addEventListener(name, (event) => {
        const handlers = this.#handlers.get(name);
        if (!handlers?.size) return;

        let payload: unknown;
        try {
          payload = JSON.parse((event as MessageEvent<string>).data);
        } catch {
          // A malformed frame must not take the stream down with it: the next event is
          // probably fine, and a dead stream is a silently stale dashboard.
          console.error(`curio: could not parse a ${name} event`);
          return;
        }
        for (const handler of handlers) handler(payload);
      });
    }

    // `hello` on connect and `ping` every 20 s are stream-level keepalives. They are
    // tolerated silently — they carry nothing and exist to keep the connection warm.
    source.addEventListener("error", () => {
      this.disconnect();
      this.#reconnect = setTimeout(() => this.connect(), RECONNECT_MS);
    });
  }

  disconnect(): void {
    this.#source?.close();
    this.#source = null;
    if (this.#reconnect) {
      clearTimeout(this.#reconnect);
      this.#reconnect = null;
    }
  }
}

/** The app's one stream (R-FE-7). Components subscribe; they never open their own. */
export const events = new EventStream();

File: crates/curio-core/src/events.rs
//! The event vocabulary.
//!
//! One in-process bus fans out to two push transports: SSE at `/api/events` for the
//! dashboard, and a WebSocket at `/ws` for the extension (D13). Both are served; neither
//! replaces the other, because they solve different problems — `EventSource` cannot send
//! headers, and an MV3 service worker dies after 30 s idle unless something keeps it warm.
//!
//! The names below are a **contract with existing clients** (R-BE-15, Inventory §3), not
//! an internal detail. Renaming one is a breaking change to the dashboard and the
//! extension simultaneously.
//!
//! Two subtleties that live in the consumers rather than here, recorded because they
//! explain the shapes:
//!
//! * `job.updated` may carry a full row **or** a partial `{id, result}` during bulk
//!   progress, and clients merge partials by id (Inventory §10.30). A progress tick for a
//!   240-item retag should not re-send the whole job row eight times a second.
//! * `item.created` is **ignored by the library grid whenever a filter or search is
//!   active** (Inventory §10.25). Prepending a card that does not match the active filter
//!   would be a lie about what the user is looking at.

use std::fmt;

/// The name of a pushed event, as it appears on the wire.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum EventName {
    /// A new item exists. Payload: the full item.
    ItemCreated,
    /// An item changed. Payload: the full item; clients replace in place.
    ItemUpdated,
    /// An item is gone. Payload: `{id}` only.
    ItemDeleted,
    /// The watcher registered a project it had not seen.
    ProjectDetected,
    /// A project's record changed — including going `missing`, which never deletes it.
    ProjectUpdated,
    /// A project record is gone, because the user asked for it. Payload: `{id}` only.
    ///
    /// The watcher never publishes this. A folder that vanishes goes `missing` and stays in
    /// the list; only an explicit "Remove" throws the record away.
    ProjectRemoved,
    /// A job was enqueued, started, progressed, or finished.
    ///
    /// Fires at every one of those boundaries, and may be partial (see module docs).
    JobUpdated,
    /// Some part of the vocabulary changed. Payload is deliberately empty `{}`: the
    /// consumer refetches rather than reconciling a diff, because merges and deletes can
    /// touch arbitrarily many rows.
    VocabularyUpdated,
}

impl EventName {
    #[must_use]
    pub fn as_str(self) -> &'static str {
        match self {
            EventName::ItemCreated => "item.created",
            EventName::ItemUpdated => "item.updated",
            EventName::ItemDeleted => "item.deleted",
            EventName::ProjectDetected => "project.detected",
            EventName::ProjectUpdated => "project.updated",
            EventName::ProjectRemoved => "project.removed",
            EventName::JobUpdated => "job.updated",
            EventName::VocabularyUpdated => "vocabulary.updated",
        }
    }

    /// Every event name, for exhaustiveness in tests and handler registries.
    #[must_use]
    pub const fn all() -> [EventName; 8] {
        [
            EventName::ItemCreated,
            EventName::ItemUpdated,
            EventName::ItemDeleted,
            EventName::ProjectDetected,
            EventName::ProjectUpdated,
            EventName::ProjectRemoved,
            EventName::JobUpdated,
            EventName::VocabularyUpdated,
        ]
    }
}

impl fmt::Display for EventName {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

/// An event ready to fan out: a name and its already-serialized payload.
///
/// The payload is `serde_json::Value` rather than a typed enum because the two transports
/// serialize identically and the domain has no stake in the difference. Keeping it opaque
/// here is what lets `curio-server` own the wire format entirely.
#[derive(Debug, Clone)]
pub struct Event {
    pub name: EventName,
    pub payload: serde_json::Value,
}

impl Event {
    #[must_use]
    pub fn new(name: EventName, payload: serde_json::Value) -> Self {
        Self { name, payload }
    }

    /// `item.deleted` carries only an id — never the item, which no longer exists.
    #[must_use]
    pub fn item_deleted(id: &str) -> Self {
        Self::new(EventName::ItemDeleted, serde_json::json!({ "id": id }))
    }

    /// `project.removed` carries only an id, for the same reason `item.deleted` does.
    #[must_use]
    pub fn project_removed(id: &str) -> Self {
        Self::new(EventName::ProjectRemoved, serde_json::json!({ "id": id }))
    }

    /// `vocabulary.updated` carries an empty object by contract; consumers refetch.
    #[must_use]
    pub fn vocabulary_updated() -> Self {
        Self::new(EventName::VocabularyUpdated, serde_json::json!({}))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn wire_names_are_the_ones_clients_already_listen_for() {
        // R-BE-15 / Inventory §3. These strings are a published contract: the dashboard's
        // handler registry and the extension both key off them, so a typo here is a
        // silent no-op in two codebases at once.
        let expected = [
            "item.created",
            "item.updated",
            "item.deleted",
            "project.detected",
            "project.updated",
            "project.removed",
            "job.updated",
            "vocabulary.updated",
        ];
        let actual: Vec<&str> = EventName::all().iter().map(|e| e.as_str()).collect();

        assert_eq!(actual, expected);
    }

    #[test]
    fn deleted_carries_only_an_id() {
        let event = Event::item_deleted("01J000000000000000000000AB");
        assert_eq!(event.payload["id"], "01J000000000000000000000AB");
        assert_eq!(
            event.payload.as_object().map(serde_json::Map::len),
            Some(1),
            "a deleted item has nothing else left to report"
        );
    }

    #[test]
    fn vocabulary_updated_is_an_empty_object_not_null() {
        // Consumers do `JSON.parse(data)` and then refetch; `null` would parse but reads
        // as "no payload was intended", and the contract says `{}`.
        assert_eq!(Event::vocabulary_updated().payload, serde_json::json!({}));
    }
}
